### PR TITLE
Document Timers

### PR DESCRIPTION
**Pull request by engineer124**
This documents the two types of times used in `z_parameter`. As timer 1 always takes priority over timer 2 when it comes to which timer to update/draw, I've named these `timer` and `subTimer` to emphasize priority, but I could see timer1 and timer2 being fine.

This also documents various states and functions related to these timers. I've gone with `environmental timers` (envTimers) for the timers pertaining to hot rooms/underwater, although I could see clash in terminology for the "kankyo" system.

There's also enough info to name the all states further (could go with `INIT`/`CENTER`/`MOVING`/`COUNTING`) which I could do. But maybe leaving the details to comments and having it as `step 1/2/3/4` for the counting up/down is enough? Open to thoughts on that

### CODE REVIEW

File: src/code/z_player_lib.c [hunk 1]
```diff
@@ -95,11 +90,16 @@ u8 sActionModelGroups[PLAYER_IA_MAX] = {
...
+typedef struct {
+    /* 0x0 */ u8 flag;
+    /* 0x2 */ u16 textId;
+} EnvTimerTextTriggerEntry; // size = 0x4
+
+EnvTimerTextTriggerEntry sEnvTimerTextTriggers[] = {
+    { ENV_TEXT_TRIGGER_HOTROOM, 0x3040 },    // PLAYER_ENV_TIMER_HOTROOM - 1
+    { ENV_TEXT_TRIGGER_UNDERWATER, 0x401D }, // PLAYER_ENV_TIMER_UNDERWATER_FLOOR - 1
+    { 0, 0x0000 },                           // PLAYER_ENV_TIMER_SWIMMING - 1
+    { ENV_TEXT_TRIGGER_UNDERWATER, 0x401D }, // PLAYER_ENV_TIMER_UNDERWATER_FREE - 1
```
Comment (Dragorn421) [inline src/code/z_player_lib.c:102]: the enum names are missing `_TYPE`

File: src/code/z_parameter.c [hunk 2]
```diff
@@ -722,15 +722,16 @@ void func_80083108(PlayState* play) {
...
                 }
 
                 gSaveContext.buttonStatus[0] = BTN_DISABLED;
 
                 for (i = 1; i < 4; i++) {
-                    if (func_8008F2F8(play) == 2) {
+                    if (Player_GetEnvTimerType(play) == PLAYER_ENV_TIMER_TYPE_UNDERWATER_FLOOR) {
                         if ((gSaveContext.equips.buttonItems[i] != ITEM_HOOKSHOT) &&
                             (gSaveContext.equips.buttonItems[i] != ITEM_LONGSHOT)) {
                             if (gSaveContext.buttonStatus[i] == BTN_ENABLED) {
```
Comment (Dragorn421) [inline src/code/z_parameter.c:737]: :spaghetti: 
There are a few spots that are totally unrelated to timers, like here for enabling hookshot underwater

Maybe there's a more generic term than "env timer type" we could find? something with "hazard" maybe?

File: src/code/z_parameter.c [hunk 3]
```diff
@@ -3072,10 +3084,10 @@ void Interface_Draw(PlayState* play) {
     static s16 spoilingItemEntrances[] = { ENTR_SPOT10_2, ENTR_SPOT07_3, ENTR_SPOT07_3 };
     static f32 D_80125B54[] = { -40.0f, -35.0f }; // unused
     static s16 D_80125B5C[] = { 91, 91 };         // unused
-    static s16 D_8015FFE0;
-    static s16 D_8015FFE2;
-    static s16 D_8015FFE4;
-    static s16 D_8015FFE6;
+    static s16 sTimerFPSTimer;
+    static s16 sTimerStateTimer;
+    static s16 sSubTimerFPSTimer;
```
Comment (Dragorn421) [inline src/code/z_parameter.c:3089]: "fps timer" -> "next second timer" ? "tick timer" ?

**Review by Dragorn421 — COMMENTED** [review]
:hourglass_flowing_sand: 

I wasn't a fan of timer / subTimer at first, but since the (non-sub) timer takes priority over subTimer in being used (processed/drawn) I guess it makes sense to use names indicating they're "coupled" :+1:

File: src/code/z_parameter.c (hunk 2 above)
Comment (engineer124) [reply src/code/z_parameter.c:737]: I actually really like the `hazard` term, so `Player_GetEnvHazardType`? I think of enemies and other obstacles as hazards too, but I think it would still need `env` or an alternative? Maybe even drop the `Type`

Comment (Dragorn421) [reply src/code/z_parameter.c:737]: Sure, EnvHazard sounds fine

it's not like this covers every hazard anyway, it's an adventure game after all, but at least the term is new and I don't think it would be better reserved for something else

File: include/z64save.h [hunk 4]
```diff
@@ -99,6 +99,49 @@ typedef struct {
     /* 0x24 */ s32 tempCollectFlags;
 } FaroresWindData; // size = 0x28
 
+typedef enum {
+    /* 0x0 */ TIMER_STATE_OFF,
+    /* 0x1 */ TIMER_STATE_ENV_0, // init env timer that counts down, total time based on health, resets on void-out, kills at 0
+    /* 0x2 */ TIMER_STATE_ENV_1, // Frozen at the screen center
+    /* 0x3 */ TIMER_STATE_ENV_2, // Move to top-left corner
+    /* 0x4 */ TIMER_STATE_ENV_3, // Begin counting down
```
[the remark this answers is not in the document]
Comment (engineer124) [reply include/z64save.h]: I went with `PREVIEW` instead of `PRESENT`

File: include/z64save.h [hunk 5]
```diff
@@ -99,6 +99,49 @@ typedef struct {
     /* 0x24 */ s32 tempCollectFlags;
 } FaroresWindData; // size = 0x28
 
+typedef enum {
+    /* 0x0 */ TIMER_STATE_OFF,
+    /* 0x1 */ TIMER_STATE_ENV_INIT, // Init env timer that counts down, total time based on health, resets on void-out, kills at 0
+    /* 0x2 */ TIMER_STATE_ENV_PREVIEW, // Display initial time, keep it fixed at the screen center
+    /* 0x3 */ TIMER_STATE_ENV_MOVE, // Move to top-left corner
+    /* 0x4 */ TIMER_STATE_ENV_TICK, // Begin counting down
```
Comment (Dragorn421) [inline include/z64save.h:107]: "begin counting down" -> just "counting down" ? (similar with others)

also, thoughts on `TIMER_STATE_ENV_` -> `TIMER_STATE_ENV_HAZARD_` ?

File: src/code/z_parameter.c [hunk 6]
```diff
@@ -3635,221 +3660,241 @@ void Interface_Draw(PlayState* play) {
...
+
+                        case SUBTIMER_STATE_DOWN_MOVE:
+                        case SUBTIMER_STATE_UP_MOVE:
                             osSyncPrintf("event_xp[1]=%d,  event_yp[1]=%d  TOTAL_EVENT_TM=%d\n",
-                                         ((void)0, gSaveContext.timerX[1]), ((void)0, gSaveContext.timerY[1]),
-                                         gSaveContext.timer2Value);
-                            svar1 = (gSaveContext.timerX[1] - 26) / D_8015FFE6;
-                            gSaveContext.timerX[1] -= svar1;
+                                         ((void)0, gSaveContext.timerX[TIMER_ID_SUB]),
+                                         ((void)0, gSaveContext.timerY[1]), gSaveContext.subTimerTime);
```
[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/z_parameter.c]: FeelsBadMan

**Review by Dragorn421 — APPROVED** [review]

File: src/code/z_player_lib.c [hunk 7]
```diff
@@ -778,35 +778,38 @@ s32 func_8008F2BC(Player* this, s32 itemAction) {
     return -1;
 }
 
-s32 func_8008F2F8(PlayState* play) {
+s32 Player_GetEnvHazard(PlayState* play) {
```
Comment (fig02) [inline src/code/z_player_lib.c:781]: we can possibly afford the full "EnvironmentHazard" here, and only use env where it needs shortening (like in enum prefixes)

File: include/z64save.h [hunk 8]
```diff
@@ -162,17 +205,17 @@ typedef struct {
...
-    /* 0x13CE */ s16 timer1State;
-    /* 0x13D0 */ s16 timer1Value;
-    /* 0x13D2 */ s16 timer2State;
-    /* 0x13D4 */ s16 timer2Value;
-    /* 0x13D6 */ s16 timerX[2];
-    /* 0x13DA */ s16 timerY[2];
+    /* 0x13CE */ s16 timerState; // See `TimerState`
+    /* 0x13D0 */ s16 timerTime; // in seconds
+    /* 0x13D2 */ s16 subTimerState; // See `SubTimerState`
+    /* 0x13D4 */ s16 subTimerTime; // in seconds
```
Comment (fig02) [inline include/z64save.h:216]: will admit "timer time" feels a bit redundant to me. is there a specific reason you wanted to change it from value like it was? think it works fine

**Comment by fig02** [comment]
jenkins failed but it seems to be an isghj machine issue. if you end up pushing more it will go again, hopefully on someone elses agent lol

File: include/z64save.h (hunk 8 above)
Comment (engineer124) [reply include/z64save.h:216]: `Value` to me felt a bit to vague and general, so replacing it with the redundant `Time` did feel slightly better to me since it at least indicates that it's the "time" instead of the many other quantities associated with the timer.

`TimerTime` doesn't flow the greatest though, so I'm not strongly opposed to going back to `TimerValue` or finding something else for the actual time left on the timer.

Comment (Dragorn421) [reply include/z64save.h:216]: In no particular order
`timerValue`
`timerTime`
`timerSeconds`
`timerCount`

I myself don't have a preference

Comment (engineer124) [reply include/z64save.h:216]: `timerSeconds` works for me

File: src/code/z_player_lib.c (hunk 7 above)
Comment (fig02) [reply src/code/z_player_lib.c]: I see you put `Environmental` first before reverting, tbh it might be better as environmental. it makes it more of an adjective instead of just "environment hazard" which could help separate it form the actual environment system

Im fine with either, but just wanted to mention it

**Review by Dragorn421 — APPROVED** [review]

**Review by fig02 — APPROVED** [review]
